### PR TITLE
Sync nes and hapi-auth-basic plugin definitions to Hapi 17.

### PR DESCRIPTION
**Pull request by saboya**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

This might change based on https://github.com/DefinitelyTyped/DefinitelyTyped/issues/23562 resolution, but it's better to merge this anyway since the types are Hapi >= 17 exclusive.

### CODE REVIEW

File: types/nes/index.d.ts [hunk 1]
```diff
@@ -23,114 +23,127 @@
...
-
-    interface ClientOptions extends NesClient.ClientOptions {}
-
-    interface ClientConnectOptions extends NesClient.ClientConnectOptions {}
-
-    interface ClientRequestOptions extends NesClient.ClientRequestOptions {}
-
-    interface ClientSubscribeFlags extends NesClient.ClientSubscribeFlags {}
+	interface SocketAuthObject {
+		isAuthenticated: boolean;
```
Comment (DanielRosenwasser) [inline types/nes/index.d.ts]: You've converted the file to mix tabs and spaces. Can you undo this please?

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]

File: types/nes/index.d.ts [hunk 2]
```diff
@@ -23,114 +23,127 @@
...
+		artifacts: any;
+	}
+
+	interface ServerBroadcastOptions {
+		user: any
+	}
+
+	interface ServerSubscriptionOptionsFilterOptions {
+		socket: Socket;
+		credentials?: any;
```
Comment (SimonSchick) [inline types/nes/index.d.ts]: See above.

File: types/nes/index.d.ts [hunk 3]
```diff
@@ -23,114 +23,127 @@
...
-    interface ClientOptions extends NesClient.ClientOptions {}
-
-    interface ClientConnectOptions extends NesClient.ClientConnectOptions {}
-
-    interface ClientRequestOptions extends NesClient.ClientRequestOptions {}
-
-    interface ClientSubscribeFlags extends NesClient.ClientSubscribeFlags {}
+	interface SocketAuthObject {
+		isAuthenticated: boolean;
+		credentials: any;
```
Comment (SimonSchick) [inline types/nes/index.d.ts]: You might want to make those interfaces so users can define these as they wish.

File: types/nes/index.d.ts [hunk 4]
```diff
@@ -23,114 +23,127 @@
...
+		artifacts: any;
+	}
+
+	interface ServerBroadcastOptions {
+		user: any
```
Comment (SimonSchick) [inline types/nes/index.d.ts]: See above.

File: types/nes/index.d.ts [hunk 5]
```diff
@@ -23,114 +23,127 @@
...
+	}
+
+	interface ServerSubscriptionOptionsAuthOptions {
+		mode?: 'required' | 'optional';
+		scope?: string | string[];
+		entity?: 'user' | 'app' | 'any';
+		index?: boolean;
+	}
+
+	export type ServerOnSubscribeWithParams = (socket: Socket, path: string, params: any) => Promise<any>;
```
Comment (SimonSchick) [inline types/nes/index.d.ts]: Ca we narrow `any`?

File: types/nes/test/broadcast-server.ts [hunk 6]
```diff
@@ -1,14 +1,14 @@
 // from https://github.com/hapijs/nes#broadcast
 
-import Hapi = require('hapi');
+import {Server} from "hapi";
 import Nes = require('nes');
 
-var server = new Hapi.Server();
+const server: Server = new Server();
```
Comment (SimonSchick) [inline types/nes/test/broadcast-server.ts]: Redundant typings imo.

File: types/nes/test/client-require.ts [hunk 7]
```diff
@@ -1,12 +1,12 @@
 // from https://github.com/hapijs/nes/#browser-client
-// When you require('nes') it loads the full module and adds a lot of extra code
+// When you require("nes") it loads the full module and adds a lot of extra code
```
Comment (SimonSchick) [inline types/nes/test/client-require.ts]: Why the random quoting change?

File: types/nes/test/nes-tests.ts [hunk 8]
```diff
@@ -1,56 +1,56 @@
-import Hapi = require('hapi');
-import Nes = require('nes');
+import {Request, ResponseToolkit, Server, ServerOptions, ServerRoute} from "hapi";
+import Nes = require("nes");
 
-var server: Hapi.Server = new Hapi.Server();
+var server: Server = new Server();
 
 server.register(Nes).then(() => {
-    // No longer need to cast to Nes.Server as Hapi.Server has been modified directly.
```
Comment (SimonSchick) [inline types/nes/test/nes-tests.ts]: Pls don't randomly change whitespaces.

**Review by SimonSchick — CHANGES_REQUESTED** [review]
Lots of unncessary style changes (please just stick to the TSLint style of the root project), some types could be exposed as interfaces to allow user extension.

**Comment by saboya** [comment]
@DanielRosenwasser @SimonSchick My apologies for the whitespace mess, my editor didn't handle a project this big well, so I deviated from my current workflow and thought I was actually "fixing" the whitespacing / quoting. I corrected everything and rebased, along with the unnecessary typings and some unnecessary imports as well.

As for the other issues, I'm gonna take a look later. My rationale behind this update was just to keep it up to date with the new Hapi bindings, so I didn't have any improvements in mind, so that definitely got overlooked.

Thanks for the input.

**Review by SimonSchick — APPROVED** [review]

**Review by DanielRosenwasser — APPROVED** [review]

**Comment by DanielRosenwasser** [comment]
Thanks!

**Pull request merged**